[16.0][IMP] pos_membership_extension: consider membership in shopping cart

flaenen opened the pull request:
Allow the sale of a product requiring membership if the membership product is in the shopping cart, even if the customer is not yet a member

legalsylvain commented:
I'm not sure to understand. After the order is confirmed, the state of the customer will allways be "Non member". That is not consistent don't you think ?

flaenen commented:
Once the order has been confirmed, the customer becomes a member of the same category as the membership product (Paid Member). This makes it possible to sell the membership at the same time as a product requiring membership.

polchampion commented:
@flaenen Tested unsuccessfully on https://lespetitescantines.test.coopiteasy.be/
At first it worked, then I changed the config (among other things I installed `l10n_fr_pos_cert_update_draft_order_line`
and now it doesn't work at all : 

![screenshot-lespetitescantines test coopiteasy be-2025 03 20-13_48_51](https://github.com/user-attachments/assets/5a9eebfa-2e42-4bb4-a426-f741c307a80d)

flaenen commented:
@polchampion On this instance, the membership product is Adhésion and not Adhésion en Cantine.  I've added it to the POS.

polchampion reviewed (approved):
Successful function test

@flaenen bien vu!

polchampion commented:
@legalsylvain did you take another look here?

remytms reviewed (approved):
LGTM

polchampion commented:
@legalsylvain when you have a moment ;-)

legalsylvain reviewed (changes requested):
Tested. Looks broken.

step to reproduce : 

- on last runboat, 
- open a session
- search "product for member"
- select it.

![image](https://github.com/user-attachments/assets/6b267641-17e5-492c-8cc4-1c081045bb59)


Result : UX Problem. the product is selected. IMO, the product could be selectable, only if a membership product has been put in the shopping cart before.

Then : 
- mark the order as paid.

Result: :no_entry: Feature problem. It is possible to sale a product reserved to membership, on an order without customer selected.

polchampion commented:
> Tested. Looks broken.
> 
> step to reproduce :
> 
>     * on last runboat,
> 
>     * open a session
> 
>     * search "product for member"
> 
>     * select it.
> 
> 
> ![image](https://private-user-images.githubusercontent.com/3407482/434753060-6b267641-17e5-492c-8cc4-1c081045bb59.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ5MDEwMzEsIm5iZiI6MTc0NDkwMDczMSwicGF0aCI6Ii8zNDA3NDgyLzQzNDc1MzA2MC02YjI2NzY0MS0xN2U1LTQ5MmMtOGNjNC0xYzA4MTA0NWJiNTkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MTdUMTQzODUxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmYyMGZmZDkxZTdiMTFhZDIxYTdjM2NlZmMwMWE5OGZhYWNlOWU0Y2JlYmE5MmFiMTkyZjkwMDZjZTJkNDMyNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.xe7OqeheJekc1DW2ZbknUsvMYzr-VbzLTh7t9L9PG0s)
> 
> Result : UX Problem. the product is selected. IMO, the product could be selectable, only if a membership product has been put in the shopping cart before.
> 
> Then :
> 
>     * mark the order as paid.
> 
> 
> Result: ⛔ Feature problem. It is possible to sale a product reserved to membership, on an order without customer selected.

@flaenen perhaps a conflict with another module?

flaenen commented:
> > Tested. Looks broken.
> > step to reproduce :
> > ```
> > * on last runboat,
> > 
> > * open a session
> > 
> > * search "product for member"
> > 
> > * select it.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ![image](https://private-user-images.githubusercontent.com/3407482/434753060-6b267641-17e5-492c-8cc4-1c081045bb59.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ5MDEwMzEsIm5iZiI6MTc0NDkwMDczMSwicGF0aCI6Ii8zNDA3NDgyLzQzNDc1MzA2MC02YjI2NzY0MS0xN2U1LTQ5MmMtOGNjNC0xYzA4MTA0NWJiNTkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MTdUMTQzODUxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmYyMGZmZDkxZTdiMTFhZDIxYTdjM2NlZmMwMWE5OGZhYWNlOWU0Y2JlYmE5MmFiMTkyZjkwMDZjZTJkNDMyNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.xe7OqeheJekc1DW2ZbknUsvMYzr-VbzLTh7t9L9PG0s)
> > Result : UX Problem. the product is selected. IMO, the product could be selectable, only if a membership product has been put in the shopping cart before.
> > Then :
> > ```
> > * mark the order as paid.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Result: ⛔ Feature problem. It is possible to sale a product reserved to membership, on an order without customer selected.
> 
> @flaenen perhaps a conflict with another module?

pos_sale_product_config_no_variant is not compatible with pos_membership_extension (no call to super in _getAddProductOptions method). This is not related to the proposed evolution.

legalsylvain commented:
Tested again.

I'm sorry. I dont understand what is working.

- possibility to sale membership without customer defined. (Doesnt make sense and should be forgotten)
- saling membership to customer doesnt make him a member...

flaenen commented:
> Tested again.
> 
> I'm sorry. I dont understand what is working.
> 
> * possibility to sale membership without customer defined. (Doesnt make sense and should be forgotten)
> * saling membership to customer doesnt make him a member...

This PR mainly concerns the code that determines whether or not a product can be added to the shopping cart. The modification is therefore very localised. It cannot be the cause of these problems.

pos_membership_extension does not require the creation of an invoice and the selection of a customer. To do this, the pos_membership module must be installed. I suggest adding it as a dependency : #1372

flaenen committed:
[IMP] pos_membership_extension: consider membership in shopping cart
The sale of a product requiring membership is allowed if the membership product
is in the shopping cart, even if the customer is not yet a member.

polchampion commented:
@legalsylvain problems you mention have a cause elsewhere or should be fixed after #1372 is merged, can you take another look then?
